walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,127 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.*;
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
+    //   Returns a collection of the TimeRanges in which mandatory attendees will be free for a given 
+    // meeting request as well as including when possible optional attendees
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Set<String> attendees = new HashSet<>(request.getAttendees());
+    Set<String> optAttendees = new HashSet<>(request.getOptionalAttendees());
+    long requestDuration = request.getDuration();
+
+
+    List<TimeRange> busyTimes = findBusyPeriods(events,attendees);
+    List<TimeRange> freeTime = findFreePeriods(busyTimes, busyTimes.size());
+    List<TimeRange> meets = parseMeetTimes(freeTime, requestDuration);
+    List<TimeRange> optBusyTimes = findBusyPeriods(events,optAttendees);
```

Comment:
At this point, why don't you use "meets" instead of starting over from events? You are going through all the time slots again, but eventually you will need to filter it for those that have all the required attendees, which you have already calculated. Just a suggestion for optimizing it, you don't need to implement it if you're not sure how.